### PR TITLE
CD-150272 Add binding for rd_kafka_seek and rd_kafka_conf_set_rebalance_cb

### PR DESCRIPTION
**Pull request by johnny-lai**
- [x] @abdollar  
- [x] @johnwu96822

### CODE REVIEW

File: lib/rdkafka/consumer.rb [hunk 1]
```diff
@@ -190,6 +215,38 @@ def lag(topic_partition_list, watermark_timeout_ms=100)
...
+        native_topic,
+        partition,
+        offset,
+        timeout_ms
+      )
+      if response != 0
+        raise Rdkafka::RdkafkaError.new(response)
+      end
+    ensure
+      if native_topic && !native_topic.null?
```
Comment (abdollar) [inline lib/rdkafka/consumer.rb]: @johnny-lai null? - is that a thing or a typo

File: examples/seek_rebalance_example.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+optparse.parse!(into: params)
+
+params[:"bootstrap-server"] ||= "localhost:9092"
+if !params.has_key?(:group) || !params.has_key?(:topic)
+  puts optparse.help
+  exit 1
+end
+
+c = Rdkafka::Config.new(
+  "bootstrap.servers": params[:"bootstrap-server"],
```
Comment (abdollar) [inline examples/seek_rebalance_example.rb]: @johnny-lai can this be a list of servers?

Comment (johnny-lai) [reply examples/seek_rebalance_example.rb]: Yes. It'd be a comma-separated list for rdkafka

File: lib/rdkafka/consumer.rb (hunk 1 above)
Comment (johnny-lai) [reply lib/rdkafka/consumer.rb]: Yes, it's `null?`. See https://github.com/coupa/rdkafka-ruby/pull/1/files#diff-59f34a2da5ed0185a7db29ef4309e72fR276 for another example

File: examples/seek_rebalance_example.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+})
+
+trap("QUIT") { c.close }
+trap("INT") { c.close }
+trap("TERM") { c.close }
+
+c.subscribe(params[:topic])
+puts "Subscribed to #{params[:topic]}"
+
+end_offset = params[:"end-offset"]
```
Comment (abdollar) [inline examples/seek_rebalance_example.rb]: @johnny-lai to_i - wondering if it should only be an int and then below you can say > 0

File: lib/rdkafka/consumer.rb (hunk 1 above)
Comment (johnny-lai) [reply lib/rdkafka/consumer.rb]: I think it's checking if the FFI pointer object contains a NULL pointer

File: examples/seek_rebalance_example.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+c.each do |message|
+  puts "Got #{message.topic}/#{message.partition}@#{message.offset} #{message.payload}"
+  if !end_offset.nil? && message.offset >= end_offset
```
Comment (abdollar) [inline examples/seek_rebalance_example.rb]: @johnny-lai end_offset > 0 - just wondering about blanks or spaces

File: examples/seek_rebalance_example.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    puts "Commiting the updated offsets immediately"
+    tpl = Rdkafka::Consumer::TopicPartitionList.new
+    tpl.add_topic_and_partitions_with_offsets(message.topic, {message.partition => 0})
+    c.commit(tpl)
+  end
+
+  # Sleep 1 second to slow down the consumption so it is easier to follow.
+  # Not needed in the real code.
+  sleep 1
```
Comment (abdollar) [inline examples/seek_rebalance_example.rb]: @johnny-lai do you want to make this configurable above for testing

File: examples/seek_rebalance_example.rb (hunk 3 above)
Comment (johnny-lai) [reply examples/seek_rebalance_example.rb]: So I was missing the `end-offset` parameter definition. That ensures `params[:"end-offset"]` will return an integer.

So this is a sample program. It's just exercising the new functions.

File: examples/seek_rebalance_example.rb (hunk 5 above)
Comment (johnny-lai) [reply examples/seek_rebalance_example.rb]: Sure. Adding `-w`

File: examples/seek_rebalance_example.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  "bootstrap.servers": params[:"bootstrap-server"],
+  "group.id": params[:group],
+).consumer(rebalance_cb: ->(consumer, err, tpl) {
+  puts "Consumer group #{params[:group]} rebalanced: #{err}"
+  case err.code
+  when :assign_partitions
+    start_offset = params[:"start-offset"]
+    unless start_offset.nil?
+      puts "Assigning partition 0 of #{params[:topic]} to #{start_offset}"
+      tpl[params[:topic]][0].offset = start_offset
```
Comment (supriya) [inline examples/seek_rebalance_example.rb]: do you need to do nil check here before calling .offset?

File: lib/rdkafka/consumer.rb (hunk 1 above)
Comment (abdollar) [reply lib/rdkafka/consumer.rb]: 👍

File: examples/seek_rebalance_example.rb (hunk 2 above)
Comment (abdollar) [reply examples/seek_rebalance_example.rb]: 👍

File: examples/seek_rebalance_example.rb (hunk 3 above)
Comment (abdollar) [reply examples/seek_rebalance_example.rb]: 👍

**Comment by abdollar** [comment]
👍

File: lib/rdkafka/consumer.rb [hunk 7]
```diff
@@ -8,6 +8,9 @@ module Rdkafka
   class Consumer
     include Enumerable
 
+    # @private
+    @rebalance_callback = nil
```
Comment (johnwu96822) [inline lib/rdkafka/consumer.rb]: This is a class level variable, different from the instance variable referenced in the methods.

**Comment by johnwu96822** [comment]
👍 One minor comment

File: lib/rdkafka/consumer.rb (hunk 7 above)
Comment (johnny-lai) [reply lib/rdkafka/consumer.rb]: Oh.. right. Hmm... I copied that from https://github.com/coupa/rdkafka-ruby/blob/master/lib/rdkafka/producer.rb#L5

Comment (johnny-lai) [reply lib/rdkafka/consumer.rb]: It is not needed. Let me remove. Thanks!

File: examples/seek_rebalance_example.rb (hunk 6 above)
Comment (johnny-lai) [reply examples/seek_rebalance_example.rb]: I've updated it to support iterating through all the partitions and topics. We need that later anyway. Thanks

**Comment by johnny-lai** [comment]
[lockit v4.9.0] merged via Lockit CLI on 2019-03-14 09:50:56.336516 -0700 PDT m=+5.852763620

**Pull request merged**